benchconnect/benchconnect/_submit.py

```diff
@@ -0,0 +1,84 @@
+from json import load
+from pathlib import Path
+
+import click
+from benchadapt.result import BenchmarkResult
+from benchclients.conbench import ConbenchClient
+
+from ._augment import augment_blob
+from ._post import post_blob
+from ._start import STATEFILE
+from .utils import ENV_VAR_HELP, load_json
+
+
+def augment_and_post_result(json: dict, client: ConbenchClient) -> None:
+    "Augment a result from the statefile and class, then post it"
+    statefile_path = Path(STATEFILE).resolve()
+
+    if not statefile_path.exists():
+        raise FileNotFoundError(
+            f"Statefile not found at {statefile_path}! Call `benchconnect start run` first."
+        )
+
+    with open(statefile_path, "r") as f:
+        abstract_result = load(f)
+
+    for result_key in abstract_result:
+        if result_key in json:
+            assert abstract_result[result_key] == json[result_key], click.BadParameter(
+                "Result metadata does not match run metadata! "
+                f"Key: {result_key} "
+                f"Run: {abstract_result[result_key]} "
+                f"Result: {json[result_key]}"
+            )
+
+        json[result_key] = abstract_result[result_key]
```

Comment:
Most of these keys will be ignored by the server except `run_id` right? I wonder if we could save a tiny bit of time by just storing `run_id` in the statefile and not checking/adding the rest of the run keys every time we post a result. Elena recently [brought up](https://github.com/conbench/conbench/pull/429#discussion_r1017272968) that it's nice to keep the time for posting BenchmarkResults down as much as possible.

Though I realize these are all speedy dict hash operations so it wouldn't save that much time haha.

Reply:
I thought about what the behavior should be here a bit, and from a least-surprise perspective, it seems like calls should fail if the result metadata conflicts with the run metadata even if it's going to be ignored; you could be posting a result for a different commit or something by accident. There are a few reasons for copying here, none huge:

1. even if the server ignores it, that metadata will get logged, which can be helpful for debugging
2. it's consistent with `augment`
3. it will stop the augmentation from overriding anything it shouldn't

We could just drop it all before posting? But the assertions feel important. Speed here seems exceedingly negligible compared to any network i/o.